### PR TITLE
Add QueryConfig to VectorFunction, Aggregate and WindowFunction

### PR DESCRIPTION
**Pull request by jinchengchenghh**
Resolves: https://github.com/facebookincubator/velox/issues/4774

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
Can you help review? Thanks! @mbasmanova

File: velox/core/QueryConfig.h [hunk 1]
```diff
@@ -192,6 +192,18 @@ class QueryConfig {
   static constexpr const char* kSparkLegacySizeOfNull =
       "spark.legacy_size_of_null";
 
+  // The default number of expected items for the runtime bloomfilter.
+  static constexpr const char* kSparkRuntimeBloomFilterExpectedNumItems =
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:196]: It would be nice to split fragment changes (add access to query config for aggregate functions) from using that functionality into separate PRs.

File: velox/exec/Aggregate.h [hunk 2]
```diff
@@ -42,6 +43,8 @@ class Aggregate {
  public:
   virtual ~Aggregate() {}
 
+  virtual void initialize(const core::QueryConfig& config) {}
```
Comment (mbasmanova) [inline velox/exec/Aggregate.h:46]: Would you document this method?

File: velox/exec/HashAggregation.cpp [hunk 3]
```diff
@@ -119,8 +119,9 @@ HashAggregation::HashAggregation(
     constantLists.push_back(constants);
   }
 
-  // Check that aggregate result type match the output type
+  // Initialize and check that aggregate result type match the output type
   for (auto i = 0; i < aggregates.size(); i++) {
+    aggregates[i]->initialize(driverCtx->queryConfig());
```
Comment (mbasmanova) [inline velox/exec/HashAggregation.cpp:124]: Would it make sense to pass config to Aggregate::create and have it pass to AggregateFunctionFactory so that config is provided at the time when function is created and not after?

File: velox/functions/lib/aggregates/tests/AggregationTestBase.cpp [hunk 4]
```diff
@@ -377,6 +395,9 @@ void AggregationTestBase::testAggregationsWithCompanion(
     }
 
     AssertQueryBuilder queryBuilder(builder.planNode(), duckDbQueryRunner_);
+    for (auto entry : config) {
```
Comment (mbasmanova) [inline velox/functions/lib/aggregates/tests/AggregationTestBase.cpp:398]: This code repeats a lot. Maybe add AssertQueryBuilder::configs(configs) method.

**Review by mbasmanova — COMMENTED** [review]
@jinchengchenghh Thank you for adding this functionality. Overall looks good. Some comments.

File: velox/core/QueryConfig.h (hunk 1 above)
Comment (jinchengchenghh) [reply velox/core/QueryConfig.h:196]: I put here because I can add unit test to make sure the functionality works well with these configs. Do I need to move it to another PR and no test for the functionality?

File: velox/exec/HashAggregation.cpp (hunk 3 above)
Comment (jinchengchenghh) [reply velox/exec/HashAggregation.cpp:124]: I try as you suggest, but it will change the code too lot, every file in folder `velox/functions/prestosql/aggregates/` and ``velox/functions/sparksql/aggregates/` will be affected, and we cannot easily get QueryConfig in https://github.com/facebookincubator/velox/blob/main/velox/exec/AggregateCompanionAdapter.cpp#L440 'registerExtractFunctionWithSuffix' unless we add QueryConfig to `VectorFunctionFactory`,  or just an empty config.

Can you check this issue?

Comment (mbasmanova) [reply velox/exec/HashAggregation.cpp:124]: CC: @kagamiori Wei, would you help look into this?

Comment (jinchengchenghh) [reply velox/exec/HashAggregation.cpp:124]: @kagamiori Do you have some idea?

Comment (jinchengchenghh) [reply velox/exec/HashAggregation.cpp:124]: Do you have some idea? @mbasmanova

Comment (mbasmanova) [reply velox/exec/HashAggregation.cpp:124]: @jinchengchenghh I think it would be cleaner to pass config to Aggregate::create and getVectorFunction. It would also make it consistent with SimpleFunctionAdapterFactory::createVectorFunction. I do appreciate that would require some refactoring and changing many files, but I think that's Ok.

File: velox/exec/AggregateCompanionAdapter.cpp [hunk 5]
```diff
@@ -435,9 +444,11 @@ bool CompanionFunctionsRegistrar::registerExtractFunctionWithSuffix(
             "Signatures whose result types are not resolvable given intermediate types should have been excluded.");
       }
 
+      // TODO: how to get config here
+      core::QueryConfig config({});
       if (auto func = getAggregateFunctionEntry(originalName)) {
         auto fn = func->factory(
-            core::AggregationNode::Step::kFinal, argTypes, resultType);
+            core::AggregationNode::Step::kFinal, argTypes, resultType, config);
```
Comment (jinchengchenghh) [inline velox/exec/AggregateCompanionAdapter.cpp:451]: Here is a special case.
It create Aggregate in VectorFunction, should we also add 'core::QueryConfig' to `VectorFunctionFactory`? @mbasmanova

Comment (mbasmanova) [reply velox/exec/AggregateCompanionAdapter.cpp:452]: Yes, let's do that.

Comment (jinchengchenghh) [reply velox/exec/AggregateCompanionAdapter.cpp:452]: Added

**Comment by jinchengchenghh** [comment]
Finished, could you help review again? This PR changes so much files, can you put it high priority to avoid conflicts? @mbasmanova

File: velox/exec/AggregateCompanionAdapter.cpp [hunk 6]
```diff
@@ -435,9 +446,10 @@ bool CompanionFunctionsRegistrar::registerExtractFunctionWithSuffix(
             "Signatures whose result types are not resolvable given intermediate types should have been excluded.");
       }
 
+      // TODO: how to get config here
```
Comment (mbasmanova) [inline velox/exec/AggregateCompanionAdapter.cpp:449]: Is this TODO still needed?

File: velox/exec/AggregateCompanionAdapter.cpp [hunk 7]
```diff
@@ -492,8 +505,9 @@ bool CompanionFunctionsRegistrar::registerExtractFunction(
     }
 
     if (auto func = getAggregateFunctionEntry(originalName)) {
+      // TODO: how to get config here?
```
Comment (mbasmanova) [inline velox/exec/AggregateCompanionAdapter.cpp]: Is this TODO still needed?

File: velox/exec/tests/HashTableTest.cpp [hunk 8]
```diff
@@ -546,12 +546,13 @@ TEST_P(HashTableTest, mixed6Sparse) {
 TEST_P(HashTableTest, clear) {
   std::vector<std::unique_ptr<VectorHasher>> keyHashers;
   keyHashers.push_back(std::make_unique<VectorHasher>(BIGINT(), 0 /*channel*/));
-
+  core::QueryConfig config({});
```
Comment (mbasmanova) [inline velox/exec/tests/HashTableTest.cpp:549]: nit: "({})" not needed

File: velox/exec/tests/utils/AssertQueryBuilder.h [hunk 9]
```diff
@@ -39,6 +39,10 @@ class AssertQueryBuilder {
   /// properties.
   AssertQueryBuilder& config(const std::string& key, const std::string& value);
 
+  /// Set multiple configuration properties
+  AssertQueryBuilder& configs(
```
Comment (mbasmanova) [inline velox/exec/tests/utils/AssertQueryBuilder.h:43]: This change seems unrelated. Better move it into a separate PR.

File: velox/expression/VectorFunction.h [hunk 10]
```diff
@@ -178,6 +178,7 @@ std::shared_ptr<const Type> resolveVectorFunction(
 /// constant input; non-constant inputs should be represented as nullptr;
 /// constant inputs must be instances of ConstantVector.
 std::shared_ptr<VectorFunction> getVectorFunction(
+    const core::QueryConfig& config,
```
Comment (mbasmanova) [inline velox/expression/VectorFunction.h:181]: Move config to be the last argument for consistency.

File: velox/functions/lib/aggregates/tests/AggregationTestBase.cpp [hunk 11]
```diff
@@ -54,32 +54,36 @@ void AggregationTestBase::testAggregations(
     const std::vector<RowVectorPtr>& data,
     const std::vector<std::string>& groupingKeys,
     const std::vector<std::string>& aggregates,
-    const std::string& duckDbSql) {
+    const std::string& duckDbSql,
+    const std::unordered_map<std::string, std::string>& config) {
```
Comment (mbasmanova) [inline velox/functions/lib/aggregates/tests/AggregationTestBase.cpp:58]: Why is this change? Seems unrelated.

File: velox/core/QueryConfig.h (hunk 1 above)
Comment (mbasmanova) [reply velox/core/QueryConfig.h:196]: This is a large PR. Having multiple logical changes combined here makes it hard to review and will be difficult to land. Let's limit this PR to just passing config to scalar, aggregate and window functions.

**Review by mbasmanova — COMMENTED** [review]
@jinchengchenghh Some comments. Please, update PR description to match the changes.

File: velox/expression/VectorFunction.h (hunk 10 above)
Comment (jinchengchenghh) [reply velox/expression/VectorFunction.h:181]: To consist with sinple funtion, I put it to be first argument.
```
      auto func = simpleFunctionEntry->createFunction()->createVectorFunction(
          config, getConstantInputs(compiledInputs));
```

```
    if (auto specialForm = getSpecialForm(
            config,
            call->name(),
            resultType,
            std::move(compiledInputs),
            trackCpuUsage)) {
      result = specialForm;
    } else if (
        auto func = getVectorFunction(
            config,
            call->name(),
            inputTypes,
            getConstantInputs(compiledInputs))) {
      result = std::make_shared<Expr>(
          resultType,
          std::move(compiledInputs),
          func,
          call->name(),
          trackCpuUsage);
    } else if (
        auto simpleFunctionEntry =
            simpleFunctions().resolveFunction(call->name(), inputTypes)) {
      VELOX_USER_CHECK(
          resultType->equivalent(*simpleFunctionEntry->type().get()),
          "Found incompatible return types for '{}' ({} vs. {}) "
          "for input types ({}).",
          call->name(),
          simpleFunctionEntry->type(),
          resultType,
          folly::join(", ", inputTypes));
      auto func = simpleFunctionEntry->createFunction()->createVectorFunction(
          config, getConstantInputs(compiledInputs));
```

Do I need to move it？

Comment (mbasmanova) [reply velox/expression/VectorFunction.h:181]: I see. Let's look at all the APIs to create functions and make sure they are consistent. Would you compile such a list so we can get a better picture?

File: velox/exec/tests/HashTableTest.cpp (hunk 8 above)
Comment (jinchengchenghh) [reply velox/exec/tests/HashTableTest.cpp:549]: If I remove it, I will meet:
```
[build] ../velox/expression/tests/ExpressionFuzzer.cpp:183:25: error: no matching function for call to ‘facebook::velox::core::QueryConfig::QueryConfig()’
[build]   183 |       core::QueryConfig config;
```

Comment (mbasmanova) [reply velox/exec/tests/HashTableTest.cpp:549]: Can you try: `core::QueryConfig config;`?

Comment (mbasmanova) [reply velox/exec/tests/HashTableTest.cpp:549]: I see. Thanks.

```
  explicit QueryConfig(
      const std::unordered_map<std::string, std::string>& values);
```

File: velox/expression/VectorFunction.h (hunk 10 above)
Comment (jinchengchenghh) [reply velox/expression/VectorFunction.h:181]: I find 3:
```
  void createWindowFunctions(
      const std::shared_ptr<const core::WindowNode>& windowNode,
      const RowTypePtr& inputType,
      const core::QueryConfig& config);

class SimpleFunctionAdapterFactory {
 public:
  virtual std::unique_ptr<VectorFunction> createVectorFunction(
      const core::QueryConfig& config,
      const std::vector<VectorPtr>& constantInputs) const = 0;
  virtual ~SimpleFunctionAdapterFactory() = default;
};

std::shared_ptr<VectorFunction> getVectorFunction(
    const core::QueryConfig& config,
    const std::string& name,
    const std::vector<TypePtr>& inputTypes,
    const std::vector<VectorPtr>& constantInputs);
```

Do I need to move WindowFunction `config` to first argument? Or move all the config to last argument?

Comment (jinchengchenghh) [reply velox/expression/VectorFunction.h:181]: If `Aggregate::create` need to aligh.
```
std::unique_ptr<Aggregate> Aggregate::create(
    const std::string& name,
    core::AggregationNode::Step step,
    const std::vector<TypePtr>& argTypes,
    const TypePtr& resultType,
    const core::QueryConfig& config) {
```

Comment (mbasmanova) [reply velox/expression/VectorFunction.h:181]: Thanks. Let's move 'config' to the end of the argument list in all these APIs.

Comment (jinchengchenghh) [reply velox/expression/VectorFunction.h]: Moved

File: velox/functions/lib/aggregates/tests/AggregationTestBase.cpp (hunk 11 above)
Comment (jinchengchenghh) [reply velox/functions/lib/aggregates/tests/AggregationTestBase.cpp]: Removed

**Comment by jinchengchenghh** [comment]
Do you have further comments? @mbasmanova

**Review by mbasmanova — APPROVED** [review]
@jinchengchenghh Looks good to me. Thanks. This change will most likely require more changes internally. Will try to get these done sometime this week. Thanks.

**Comment by mbasmanova** [comment]
CI is red. Please, take a look.
